Add API groups explanation to golang tutorial (#5842)
operator-framework/operator-sdk: website/content/en/docs/building-operators/golang/tutorial.md

# so all API groups will be <group>.example.com
operator-sdk init --domain example.com --repo github.com/example/memcached-operator
```

> `--domain` will be used as the prefix of the API group your custom resources will be created in.
> API groups are a mechanism to group portions of the Kubernetes API. You're probably already familiar with
> some of the core Kubernetes API groups, such as `apps` or `rbac.authorization.k8s.io`. You should name your
> domain to group your resource types in meaningful group(s), both for ease of understanding and because these
> groups determine how access can be controlled to your resource types using RBAC. For more information, see [the core Kubernetes docs](https://kubernetes.io/docs/reference/using-api/#api-groups) and [the Kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/gvks.html).

Reviewer: Just to clarify:

>  You should name your
> domain to group your resource types in meaningful group(s), both for ease of understanding and because these
> groups determine how access can be controlled to your resource types using RBAC.
> [@jberkhahn](https://kubernetes.slack.com/team/U7Q21LH2S)

 I do not agree with the above text added in the PR. It is not a correct info and that is the reason I do not agree with your PR.

```suggestion
> Note: By following the [Operators pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) you will check in this tutorial that you will be able to extend the k8s api and create your own APIs. These APIs will be grouped by via the [ApiGroup](https://kubernetes.io/docs/reference/using-api/#api-groups). Therefore, you will see that the value informed via the `--domain` flag will be the subdomain of your APIs. In this case, you probably want to look for something meaningful to your organization and/or project.
```

> domain to group your resource types in meaningful group(s), both for ease of understanding and because these
> groups determine how access can be controlled to your resource types using RBAC. For more information, see [the core Kubernetes docs](https://kubernetes.io/docs/reference/using-api/#api-groups) and [the Kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/gvks.html).

Reviewer 2: @jberkhahn IMO mentioning RBAC isn't necessary since that's not the only thing for which we have `groups`. 
Rest looks good, we can merge it. 

```suggestion
domain to group your resource types in meaningful group(s). For more information, see [the core Kubernetes docs](https://kubernetes.io/docs/reference/using-api/#api-groups) and [the Kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/gvks.html).
```

# so all API groups will be <group>.example.com
operator-sdk init --domain example.com --repo github.com/example/memcached-operator
```

> API groups are a mechanism to group portions of the Kubernetes API. You're probably already familiar with

Reviewer: My ihmo about this one is: 

> API groups are a mechanism to group portions of the Kubernetes API. You're probably already familiar with some of the core Kubernetes API groups, such as `apps` or `rbac.authorization.k8s.io`. You should name your domain to group your resource types in meaningful group(s), both for ease of understanding and because these groups determine how access can be controlled to your resource types using RBAC. 

I am not sure if I can go along with this explanation. We do not create meaningful names for we are able to use RBAC. 

Note that k8s is a REST API such as any other. When we extend the API and create a CRD we are creating endpoints. Therefore, RBAC is a mechanism that defined the permissions to request GET/POST/PATCH/PUT these endpoints.

> For more information, see [the core Kubernetes docs](https://kubernetes.io/docs/reference/using-api/#api-groups) and [the Kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/gvks.html).

So, I do not think that the above explanation addresses the question:
- Why do I need to inform a domain? How the domain is used? 
- Why I can have only one domain with the tool for now?
- What happens if I want to reconcile (create controllers) using the subcommand responsible for ( create api ) using APIs from another domain like external types ( owned by other projects ) or (core types)

Author: RBAC is relevant because that's the primary way permissions to access an operator's API is going to be controlled, so the shape and contents of your domains determines what granularity of access you can have to those APIs. I assume an operator developer doesn't really care that in reality this is making REST endpoints exist on a webserver somewhere on the cluster because they don't interact with that abstraction in any way. 

I don't understand out the multi-domain stuff is relevant here at all, at least until it's written and kubebuilder and we bump the dependency here.

Reviewer: I am totally ok with we change the text but I think would be great if we came back to the original motivation and ensure that we are properly adding meaningful info here. Therefore, I do not think that we should add `domain to group your resource types in meaningful group(s),` at all. That is not accurate.

> API groups are are a mechanism to group portions of the Kubernetes API. You're probably already familiar with

Reviewer 3: API groups are a mechanism to group portions of the Kubernetes API. You're probably already familiar with

> API groups are are a mechanism to group portions of the Kubernetes API. You're probably already familiar with

Reviewer 3: API groups are a mechanism to group portions of the Kubernetes API. You're probably already familiar with

Reviewer: WDYT about nere we just link the doc: https://book.kubebuilder.io/cronjob-tutorial/gvks.html
Suggestion:

**Note** If you are not familiarized with Kubebernets APIs please ensure that you check [Groups and Versions and Kinds, oh my!](https://book.kubebuilder.io/cronjob-tutorial/gvks.html).

Author: fixed

some of the core Kubernetes API groups, such as `apps` or `rbac.authorization.k8s.io`. You should name your
domain to group your resource types in meaningful group(s), both for ease of understanding and because these

> groups determine how access can be controlled to your resource types using RBAC. For more information, see [the core Kubernetes docs](https://kubernetes.io/docs/reference/using-api/#api-groups).

Reviewer: The RBAC is generated via the markers we have a part/section on the tutorial that only speaks about it. Ditto so, should not any extra info about that fits better there? 

Also, I am not sure if I can go along with `domain to group your resource types in meaningful group(s), both for ease of understanding and because these
groups determine how access can be controlled to your resource types using RBA`

- It will create the endpoints for the k8s API
- The RBAC give permissions to HTTP protocol permissions ( PATCH/DELETE/CRETE ) to these endpoints 
- In the project scaffold by default, we have RBAC permissions that are generated to allow the Operator/manager to perform these operations ( verbs ) for what is required
But we also scaffold permissions to help cluster admins create groups to allow view/edit each GKV scaffold.

Reviewer: I am closing this one and I hope that I could clarify all points here: https://github.com/operator-framework/operator-sdk/pull/5842#discussion_r930702102

> API groups are a mechanism to group portions of the Kubernetes API. You're probably already familiar with
> some of the core Kubernetes API groups, such as `apps` or `rbac.authorization.k8s.io`. You should name your
> domain to group your resource types in meaningful group(s), both for ease of understanding and because these
> groups determine how access can be controlled to your resource types using RBAC. For more information, see [the core Kubernetes docs](https://kubernetes.io/docs/reference/using-api/#api-groups) and [the Kubebuilder docs](https://book.kubebuilder.io/cronjob-tutorial/gvks.html).

Reviewer: ```suggestion
**Note:** Currently, the tool does not support work with multi-domains to create the APIs by default. Therefore, you will check below that when you use the subCommand `create api` you will not need to inform the domain anymore. You will only inform the group, the KInd and the Version that should be used. Therefore, all APIs owned by the project will be generated using the domain informed when you create the project.
```

Reviewer: @jberkhahn I understand that the goal is to clarify that the APIs will be generated using the domain informed here, so I thought of something like the above. 

Also, note that the tool SHOULD be able to work/handle external APIs and we should be able to use create API to for example scaffold a controller for the core apis or apis created in another project which will have another domain. That is a limitation on the tool that we need to address see: https://github.com/kubernetes-sigs/kubebuilder/issues/1999

c/c @laxmikantbpandhare @asmacdo

Author: I don't think we should have a thing about a potential feature in a repo for another project linked in the Operator SDK docs. I think at most a blurb stating that multiple domains are not currently supported, but I don't know how I feel about explaining non-supported features, as it's not possible to enumerate everything we don't support.

Reviewer: @jberkhahn,

What motivates you to want to add it to the docs?
Someone who would like to create APIs that have another domain
Someone, that wants to create controllers/api in the project to reconcile external/core types 

Am I right? What is the motivation to add this note in this doc? 

> I don't think we should have a thing about a potential feature in a repo for another project linked in the Operator SDK docs

SDK is Kubebuilder. SDK does not implement anything about what is scaffold by default. SDK only implement its helpers and specific features and we clarify it in other places as we link Kubebuilder docs/issues and etc in many other places.  

Therefore, it is an open source project and links the issue enables the community ( those that are looking for the feature) to contribute or vote and let us know how much that is or not important for them.

Reviewer: To address the detailed info via FAQ I created an issue as discussed : https://github.com/operator-framework/operator-sdk/issues/5979
